plugins/irma-client/state-client.js

```diff
@@ -0,0 +1,283 @@
+
+    // Skip the request when the pairing method is correctly set already.
+    if (shouldBeEnabled === this._pairingEnabled) return Promise.resolve();
+
+    if (!this._mappings.frontendAuth) {
+      console.warn('Pairing cannot be enabled, because no frontendAuth token is provided. This might be a security risk.')
+      return Promise.resolve();
+    }
+    this._pairingEnabled = shouldBeEnabled;
+
+    // If pairing should be enabled, parse the pairing options struct.
+    let options = shouldBeEnabled
+      ? { pairingMethod: this._options.state.pairing.pairingMethod }
+      : { pairingMethod: 'none' };
+    return this._updateFrontendOptions(options)
+      .catch(err => {
+        if ( this._options.debugging )
+          console.error('Error received while updating pairing state:', err);
+        this._handleNoSuccess('fail', err);
+      });
+  }
+
+  _pairingCompleted() {
+    let delay = new Promise(resolve => setTimeout(resolve, this._options.state.pairing.minCheckingDelay));
+    let url = this._options.state.pairing.completedUrl(this._mappings);
+
+    fetch(url, {
+      method: 'POST',
+      headers: {'Authorization': this._mappings.frontendAuth}
+    })
+      .finally(() => delay)
+      .catch(err => {
+        if ( this._options.debugging )
+          console.error('Error received while completing pairing:', err);
+        this._handleNoSuccess('fail', err);
+      });
+  }
+
+  _updateFrontendOptions(options) {
+    if (!this._mappings.frontendAuth) return Promise.reject(new Error('frontendAuth token was not supplied'));
+
+    let req = {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        'Authorization': this._mappings.frontendAuth
+      },
+      body: JSON.stringify({
+        '@context': this._options.state.frontendOptions.requestContext,
+        ...options
+      })
+    };
+    return fetch(this._options.state.frontendOptions.url(this._mappings), req)
+      .then(r => r.json())
+      .then(options => this._frontendOptions = options);
+  }
+
+  _getMobileUrl(sessionPtr) {
+    const json = JSON.stringify(sessionPtr);
+    const universalLink = `https://irma.app/-/session#${encodeURIComponent(json)}`;
+    switch (this._userAgent) {
+      case 'Android':
+        // Universal links are not stable in Android webviews and custom tabs, so always use intent links.
+        let intent = `Intent;package=org.irmacard.cardemu;scheme=irma;l.timestamp=${Date.now()}`;
+        let fallback = `S.browser_fallback_url=${encodeURIComponent(universalLink)}`;
+        return `intent://qr/json/${encodeURIComponent(json)}#${intent};${fallback};end`;
+      case 'iOS':
+        return universalLink;
+      default:
+        throw new Error('Device type is not supported.');
+    }
+  }
+
+  _determineFlow() {
+    this._userAgent = userAgent();
+    // A new transition cannot be started within stateChange, so add call to javascript event loop.
+    Promise.resolve().then(() => {
```

Comment:
I'm still not really satisfied by the current solution that all calls to `transition` within a `stateChange` have to be wrapped in a Promise.

The problem is as follows: the current implementation of `transition` and `finalTransition` are fully non-async. This means that `transition` will block until the state change is fully processed. This is however a problem when calling this from `stateChange`, because `transition` will call `stateChange` itself. This means that the order of change events can be messed up. 

For example, lets say that the first plugin immediately triggers `abort` when getting in the `Cancelled` state, then a second plugin will first see the `Aborted` state change (because this one is called in the blocking `transition` call of the first plugin) and then see the `Cancelled` state change (when the `transition` call unblocks). This can lead to all kinds of ugly race conditions.

I considered making `transition` and `finalTransition` async (so with Promises), such that they do not block anymore. Then this issue is solved. Since JavaScript only uses one event loop, async calls are prosponed and added to the event queue. Therefore the order of state changes remains intact.

However, this is then again a disadvantage for the calls to `currentState()` or `isValidTransition()` of the state machine. If you for example check whether a certain transition is valid, you can never be sure that the transition will be valid when the transition is actually done. Concurrent `transition` calls can get in between and change the state before the transition we are planning can.

This means that when `transition` and `finalTransition` become async (which would be nice), then `currentState()`, `isEndState()` and `isValidTransition()` calls should be forbidden within `stateChange`. I doubt whether that is really an improvement. Of course for `currentState()` and `isEndState()` are not an issue, because when handling a state change event, you know what the next state is going to be. You don't need them. This means that only the `isValidTransition()` call would become impossible then. Instead, a developer should catch the `reject()` of the `transition` call then to detect whether a transition is invalid.

Would this be better or can someone maybe think of a better pattern to use here?

Reply:
I see no problem with adding the transition (in order of the plugins) to the event queue. I even think it's good practice.

I still see one problem though. For example:
Some plugin wants to monitor the state and perhaps change the normal order in which the next state is then triggered, right? Your solution makes sure that the order of transitions remains the same, but the original transition is still queue'd after the plugin's? The transition can then not be completed because, for example, the transition is not valid anymore.

I think a viable pattern here is callback chaining, which can be invoked at certain times, e.g in this case, when leaving a state. Again, these callbacks might be over-engineering, but they give plugin developers a lot of room. The callback can choose to either call the next callback in the chain or return altogether, cancelling the original transition.

Reply:
Fixed in #40